File: legacy-source/pharma-track/app/Services/CommandeFournisseurService.php
<?php

namespace App\Services;

use App\Models\FournisseurMedicament;
use App\Models\CommandeFournisseur;
use App\Models\CommandeFournisseurLigne;
use App\Models\Alerte;
use Illuminate\Support\Facades\Log;

class CommandeFournisseurService
{
    /**
     * Vérifier la disponibilité d'un médicament chez un fournisseur
     * Prend en compte le stock minimum (stock réservé)
     */
    public function verifierDisponibilite($fournisseurMedicamentId, $quantiteDemandee)
    {
        $fm = FournisseurMedicament::with('fournisseur', 'medicament')->find($fournisseurMedicamentId);
        
        if (!$fm) {
            return [
                'disponible' => false,
                'raison' => 'Produit non trouvé'
            ];
        }
        
        if (!$fm->disponible) {
            return [
                'disponible' => false,
                'raison' => 'Produit indisponible chez ce fournisseur'
            ];
        }
        
        // Récupérer le stock minimum (10 par défaut)
        $stockMinimum = $fm->stock_minimum ?? 10;
        $stockDisponiblePourVente = $fm->stock_disponible - $stockMinimum;
        
        // Stock minimum non atteint
        if ($stockDisponiblePourVente < 0) {
            return [
                'disponible' => false,
                'type' => 'stock_minimum',
                'raison' => 'Stock minimum requis non atteint. Stock minimum: ' . $stockMinimum . ' unités. Stock actuel: ' . $fm->stock_disponible,
                'stock_actuel' => $fm->stock_disponible,
                'stock_minimum' => $stockMinimum,
                'medicament_nom' => $fm->medicament->nom_commercial_fr ?? 'N/A'
            ];
        }
        
        // Commande complète (dans la limite du stock disponible pour vente)
        if ($stockDisponiblePourVente >= $quantiteDemandee) {
            return [
                'disponible' => true,
                'type' => 'complet',
                'quantite_disponible' => $stockDisponiblePourVente,
                'fournisseur' => $fm->fournisseur,
                'prix' => $fm->prix_achat,
                'medicament_nom' => $fm->medicament->nom_commercial_fr ?? 'N/A',
                'stock_minimum' => $stockMinimum,
                'stock_actuel' => $fm->stock_disponible
            ];
        }
        
        // Commande partielle (stock disponible pour vente insuffisant mais > 0)
        if ($stockDisponiblePourVente > 0) {
            return [
                'disponible' => true,
                'type' => 'partiel',
                'quantite_disponible' => $stockDisponiblePourVente,
                'quantite_manquante' => $quantiteDemandee - $stockDisponiblePourVente,
                'fournisseur' => $fm->fournisseur,
                'prix' => $fm->prix_achat,
                'medicament_nom' => $fm->medicament->nom_commercial_fr ?? 'N/A',
                'stock_minimum' => $stockMinimum,
                'stock_actuel' => $fm->stock_disponible
            ];
        }
        
        // Rupture totale (plus rien à vendre)
        return [
            'disponible' => false,
            'type' => 'indisponible',
            'raison' => 'Stock insuffisant. Stock minimum requis: ' . $stockMinimum . ' unités. Stock actuel: ' . $fm->stock_disponible,
            'stock_actuel' => $fm->stock_disponible,
            'stock_minimum' => $stockMinimum,
            'medicament_nom' => $fm->medicament->nom_commercial_fr ?? 'N/A'
        ];
    }
    
    /**
     * Trouver des fournisseurs alternatifs pour un médicament
     */
    public function trouverAlternatifs($medicamentId, $quantiteDemandee)
    {
        $alternatifs = FournisseurMedicament::where('medicament_id', $medicamentId)
            ->where('disponible', true)
            ->whereRaw('stock_disponible - IFNULL(stock_minimum, 10) >= ?', [$quantiteDemandee])
            ->with('fournisseur', 'medicament')
            ->orderBy('prix_achat', 'asc')
            ->get();
        
        return $alternatifs;
    }
    
    /**
     * Créer une commande et mettre à jour le stock
     */
    public function creerCommande($fournisseurMedicamentId, $quantite, $pharmacieId = null)
    {
        $fm = FournisseurMedicament::with('fournisseur', 'medicament')->find($fournisseurMedicamentId);
        $verification = $this->verifierDisponibilite($fournisseurMedicamentId, $quantite);
        
        if (!$verification['disponible']) {
            return [
                'success' => false,
                'message' => $verification['raison'],
                'stock_actuel' => $verification['stock_actuel'] ?? 0,
                'stock_minimum' => $verification['stock_minimum'] ?? 10,
                'alternatifs' => $this->trouverAlternatifs($fm->medicament_id, $quantite)
            ];
        }
        
        $stockMinimum = $fm->stock_minimum ?? 10;
        $quantiteDisponibleVente = $fm->stock_disponible - $stockMinimum;
        $quantiteCommandee = min($quantite, $quantiteDisponibleVente);
        $quantiteManquante = $quantite - $quantiteCommandee;
        $stockAvant = $fm->stock_disponible;
        $nomMedicament = $fm->medicament->nom_commercial_fr ?? 'Médicament';
        
        // Créer la commande
        $commande = CommandeFournisseur::create([
            'numero_commande' => 'CMD-' . date('YmdHis') . '-' . rand(100, 999),
            'fournisseur_id' => $fm->fournisseur_id,
            'pharmacie_id' => $pharmacieId,
            'user_id' => auth()->id(),
            'date_commande' => now(),
            'statut' => $quantiteManquante > 0 ? 'partiel' : 'confirmee',
            'total_ht' => $quantiteCommandee * $fm->prix_achat,
            'total_ttc' => $quantiteCommandee * $fm->prix_achat,
            'notes' => $quantiteManquante > 0 ? "Commande partielle. Manque {$quantiteManquante} unités. Stock minimum réservé: {$stockMinimum}" : null
        ]);
        
        // Ajouter la ligne de commande
        CommandeFournisseurLigne::create([
            'commande_id' => $commande->id,
            'medicament_id' => $fm->medicament_id,
            'quantite' => $quantiteCommandee,
            'quantite_demandee' => $quantite,
            'stock_avant' => $stockAvant,
            'prix_unitaire' => $fm->prix_achat,
            'total_ligne' => $quantiteCommandee * $fm->prix_achat
        ]);
        
        // Mettre à jour le stock (garder le stock minimum)
        $nouveauStock = $stockMinimum;
        $fm->stock_disponible = $nouveauStock;
        $fm->save();
        
        Log::info('Commande créée avec stock minimum', [
            'commande_id' => $commande->id,
            'quantite_commandee' => $quantiteCommandee,
            'quantite_manquante' => $quantiteManquante,
            'stock_avant' => $stockAvant,
            'stock_apres' => $nouveauStock,
            'stock_minimum' => $stockMinimum
        ]);
        
        // Vérifier l'état du stock et créer l'alerte appropriée
        $this->verifierEtatStock($fm, $nomMedicament);
        
        return [
            'success' => true,
            'type' => $quantiteManquante > 0 ? 'partiel' : 'complet',
            'commande' => $commande,
            'quantite_commandee' => $quantiteCommandee,
            'quantite_manquante' => $quantiteManquante,
            'stock_avant' => $stockAvant,
            'stock_apres' => $nouveauStock,
            'stock_minimum' => $stockMinimum,
            'message' => $quantiteManquante > 0 
                ? "Commande partielle : {$quantiteCommandee} unités commandées. {$quantiteManquante} unités en attente. Stock minimum réservé: {$stockMinimum}"
                : "Commande complète : {$quantiteCommandee} unités commandées. Stock minimum réservé: {$stockMinimum}"
        ];
    }
    
    /**
     * Vérifier l'état du stock et créer une alerte appropriée (uniquement pour le fournisseur)
     */
    private function verifierEtatStock($fournisseurMedicament, $nomMedicament)
    {
        $stockRestant = $fournisseurMedicament->stock_disponible;
        $stockMinimum = $fournisseurMedicament->stock_minimum ?? 10;
        $seuilAlerte = $fournisseurMedicament->seuil_reapprovisionnement ?? 20;
        
        // Stock au niveau du minimum (alerte pour le fournisseur uniquement)
        if ($stockRestant <= $stockMinimum) {
            Alerte::create([
                'type' => 'stock_minimum_atteint',
                'niveau' => 'eleve',
                'message' => "⚠️ STOCK MINIMUM ATTEINT : {$nomMedicament}. Stock restant: {$stockRestant} unités (minimum: {$stockMinimum}). Veuillez réapprovisionner.",
                'est_lue' => false,
                'donnees_concernees' => [
                    'fournisseur_id' => $fournisseurMedicament->fournisseur_id,
                    'medicament_id' => $fournisseurMedicament->medicament_id,
                    'medicament_nom' => $nomMedicament,
                    'stock_restant' => $stockRestant,
                    'stock_minimum' => $stockMinimum,
                    'for_fournisseur' => true
                ]
            ]);
            Log::warning('Stock minimum atteint', [
                'medicament' => $nomMedicament,
                'stock_restant' => $stockRestant,
                'stock_minimum' => $stockMinimum
            ]);
        }
        // Stock faible (proche du minimum)
        elseif ($stockRestant <= $seuilAlerte && $stockRestant > $stockMinimum) {
            Alerte::create([
                'type' => 'stock_faible',
                'niveau' => 'moyen',
                'message' => "📉 STOCK FAIBLE : {$nomMedicament}. Stock restant: {$stockRestant} unités. Seuil d'alerte: {$seuilAlerte}",
                'est_lue' => false,
                'donnees_concernees' => [
                    'fournisseur_id' => $fournisseurMedicament->fournisseur_id,
                    'medicament_id' => $fournisseurMedicament->medicament_id,
                    'medicament_nom' => $nomMedicament,
                    'stock_restant' => $stockRestant,
                    'seuil_alerte' => $seuilAlerte,
                    'for_fournisseur' => true
                ]
            ]);
            Log::info('Stock faible', [
                'medicament' => $nomMedicament,
                'stock_restant' => $stockRestant,
                'seuil_alerte' => $seuilAlerte
            ]);
        }
    }
}

File: backend/src/main/java/com/pharmatrack/fournisseur/service/CommandeFournisseurService.java
package com.pharmatrack.fournisseur.service;

import com.pharmatrack.auth.entity.User;
import com.pharmatrack.auth.repository.UserRepository;
import com.pharmatrack.catalog.entity.Alerte;
import com.pharmatrack.catalog.entity.AlerteNiveau;
import com.pharmatrack.catalog.entity.AlerteType;
import com.pharmatrack.catalog.entity.Medicament;
import com.pharmatrack.catalog.repository.AlerteRepository;
import com.pharmatrack.catalog.repository.MedicamentRepository;
import com.pharmatrack.common.api.PageUtils;
import com.pharmatrack.common.api.PagedResponse;
import com.pharmatrack.common.error.ResourceNotFoundException;
import com.pharmatrack.fournisseur.dto.CommandeResponse;
import com.pharmatrack.fournisseur.dto.CommandeResult;
import com.pharmatrack.fournisseur.dto.DisponibiliteResponse;
import com.pharmatrack.fournisseur.dto.FournisseurMedicamentResponse;
import com.pharmatrack.fournisseur.dto.MedicamentSelectionResponse;
import com.pharmatrack.fournisseur.entity.CommandeFournisseur;
import com.pharmatrack.fournisseur.entity.CommandeFournisseurLigne;
import com.pharmatrack.fournisseur.entity.CommandeStatut;
import com.pharmatrack.fournisseur.entity.FournisseurMedicament;
import com.pharmatrack.fournisseur.mapper.FournisseurMapper;
import com.pharmatrack.fournisseur.repository.CommandeFournisseurRepository;
import com.pharmatrack.fournisseur.repository.FournisseurMedicamentRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.util.StringUtils;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;
import java.util.concurrent.ThreadLocalRandom;
import java.util.function.Function;

/**
 * Pharmacy-side purchase order flow: pick a medicament, list its suppliers
 * (cheapest first), verify availability, place the order (full or partial) and
 * consult an order. Mirror of the legacy
 * {@code CommandeFournisseurService}+{@code CommandeFournisseurController}.
 */
@Service
@Transactional
public class CommandeFournisseurService {

    private static final Logger log = LoggerFactory.getLogger(CommandeFournisseurService.class);

    private final MedicamentRepository medicamentRepository;
    private final FournisseurMedicamentRepository fournisseurMedicamentRepository;
    private final CommandeFournisseurRepository commandeRepository;
    private final AlerteRepository alerteRepository;
    private final UserRepository userRepository;
    private final FournisseurMapper mapper;

    public CommandeFournisseurService(MedicamentRepository medicamentRepository,
                                      FournisseurMedicamentRepository fournisseurMedicamentRepository,
                                      CommandeFournisseurRepository commandeRepository,
                                      AlerteRepository alerteRepository,
                                      UserRepository userRepository,
                                      FournisseurMapper mapper) {
        this.medicamentRepository = medicamentRepository;
        this.fournisseurMedicamentRepository = fournisseurMedicamentRepository;
        this.commandeRepository = commandeRepository;
        this.alerteRepository = alerteRepository;
        this.userRepository = userRepository;
        this.mapper = mapper;
    }

    @Transactional(readOnly = true)
    public PagedResponse<MedicamentSelectionResponse> selection(String search,
                                                                Integer page, Integer size) {
        Pageable pageable = PageRequest.of(PageUtils.clampPage(page), PageUtils.clampSize(size),
                Sort.by(Sort.Direction.ASC, "nomCommercialFr"));
        String term = search == null ? "" : search.trim();
        Page<Medicament> result = StringUtils.hasText(term)
                ? medicamentRepository
                        .findByCodeCipContainingIgnoreCaseOrNomCommercialFrContainingIgnoreCaseOrNomCommercialArContainingIgnoreCaseOrDciContainingIgnoreCase(
                                term, term, term, term, pageable)
                : medicamentRepository.findAll(pageable);

        List<MedicamentSelectionResponse> content = result.getContent().stream()
                .map(mapper::toSelection)
                .toList();
        return new PagedResponse<>(content, result.getNumber(), result.getSize(),
                result.getTotalElements(), result.getTotalPages());
    }

    @Transactional(readOnly = true)
    public List<FournisseurMedicamentResponse> fournisseursPourMedicament(UUID medicamentId) {
        return fournisseurMedicamentRepository
                .findByMedicamentIdAndDisponibleTrueOrderByPrixAchatAsc(medicamentId).stream()
                .map(mapper::toFmResponse)
                .toList();
    }

    @Transactional(readOnly = true)
    public DisponibiliteResponse verifierDisponibilite(UUID fournisseurMedicamentId, int quantite) {
        FournisseurMedicament fm = fournisseurMedicamentRepository
                .findById(fournisseurMedicamentId).orElse(null);
        if (fm == null) {
            return new DisponibiliteResponse(false, "produit_non_trouve",
                    "Produit non trouvé", null, null, null, null, null, null, null, null);
        }
        if (!fm.isDisponible()) {
            return new DisponibiliteResponse(false, "indisponible_fournisseur",
                    "Produit indisponible chez ce fournisseur", null, null,
                    fm.getStockMinimum(), fm.getStockDisponible(), fm.getPrixAchat(),
                    nom(fm), fm.getFournisseur().getId(), fm.getFournisseur().getRaisonSociale());
        }
        int stockMinimum = fm.getStockMinimum();
        int stockDisponiblePourVente = fm.getStockDisponible() - stockMinimum;

        if (stockDisponiblePourVente < 0) {
            return new DisponibiliteResponse(false, "stock_minimum",
                    "Stock minimum requis non atteint. Stock minimum: " + stockMinimum
                            + " unités. Stock actuel: " + fm.getStockDisponible(),
                    null, null, stockMinimum, fm.getStockDisponible(), fm.getPrixAchat(),
                    nom(fm), fm.getFournisseur().getId(), fm.getFournisseur().getRaisonSociale());
        }
        if (stockDisponiblePourVente >= quantite) {
            return new DisponibiliteResponse(true, "complet", null,
                    stockDisponiblePourVente, 0, stockMinimum, fm.getStockDisponible(),
                    fm.getPrixAchat(), nom(fm), fm.getFournisseur().getId(),
                    fm.getFournisseur().getRaisonSociale());
        }
        if (stockDisponiblePourVente > 0) {
            return new DisponibiliteResponse(true, "partiel", null,
                    stockDisponiblePourVente, quantite - stockDisponiblePourVente,
                    stockMinimum, fm.getStockDisponible(), fm.getPrixAchat(), nom(fm),
                    fm.getFournisseur().getId(), fm.getFournisseur().getRaisonSociale());
        }
        return new DisponibiliteResponse(false, "indisponible",
                "Stock insuffisant. Stock minimum requis: " + stockMinimum
                        + " unités. Stock actuel: " + fm.getStockDisponible(),
                null, null, stockMinimum, fm.getStockDisponible(), fm.getPrixAchat(),
                nom(fm), fm.getFournisseur().getId(), fm.getFournisseur().getRaisonSociale());
    }

    public CommandeResult passerCommande(UUID fournisseurMedicamentId, int quantite, UUID userId) {
        FournisseurMedicament fm = fournisseurMedicamentRepository
                .findById(fournisseurMedicamentId)
                .orElseThrow(() -> new ResourceNotFoundException(
                        "fournisseur_medicament", fournisseurMedicamentId));

        DisponibiliteResponse verification = verifierDisponibilite(fournisseurMedicamentId, quantite);
        if (!verification.disponible()) {
            return new CommandeResult(
                    false,
                    verification.type(),
                    null,
                    0, 0,
                    verification.stockActuel() == null ? 0 : verification.stockActuel(),
                    0,
                    verification.stockMinimum() == null ? 10 : verification.stockMinimum(),
                    verification.raison(),
                    alternatifs(fm.getMedicament().getId(), quantite));
        }

        int stockMinimum = fm.getStockMinimum();
        int quantiteDisponibleVente = fm.getStockDisponible() - stockMinimum;
        int quantiteCommandee = Math.min(quantite, quantiteDisponibleVente);
        int quantiteManquante = quantite - quantiteCommandee;
        int stockAvant = fm.getStockDisponible();
        String nomMedicament = nom(fm);

        BigDecimal prixUnitaire = fm.getPrixAchat();
        BigDecimal total = prixUnitaire.multiply(BigDecimal.valueOf(quantiteCommandee));

        CommandeFournisseur commande = new CommandeFournisseur();
        commande.setNumeroCommande(genererNumero());
        commande.setFournisseur(fm.getFournisseur());
        commande.setPharmacieId(pharmacieIdOf(userId));
        commande.setUserId(userId);
        commande.setDateCommande(LocalDate.now());
        commande.setStatut(quantiteManquante > 0 ? CommandeStatut.partiel : CommandeStatut.confirmee);
        commande.setTotalHt(total);
        commande.setTotalTtc(total);
        commande.setNotes(quantiteManquante > 0
                ? "Commande partielle. Manque " + quantiteManquante
                        + " unités. Stock minimum réservé: " + stockMinimum
                : null);

        CommandeFournisseurLigne ligne = new CommandeFournisseurLigne();
        ligne.setCommande(commande);
        ligne.setMedicament(fm.getMedicament());
        ligne.setQuantite(quantiteCommandee);
        ligne.setQuantiteDemandee(quantite);
        ligne.setStockAvant(stockAvant);
        ligne.setPrixUnitaire(prixUnitaire);
        ligne.setTotalLigne(total);
        commande.getLignes().add(ligne);

        commandeRepository.save(commande);

        int nouveauStock = stockMinimum;
        fm.setStockDisponible(nouveauStock);
        fournisseurMedicamentRepository.save(fm);

        log.info("Commande créée avec stock minimum [commande={}, quantite_commandee={}, "
                        + "quantite_manquante={}, stock_avant={}, stock_apres={}, stock_minimum={}]",
                commande.getNumeroCommande(), quantiteCommandee, quantiteManquante,
                stockAvant, nouveauStock, stockMinimum);

        verifierEtatStock(fm, nomMedicament);

        String message = quantiteManquante > 0
                ? "Commande partielle : " + quantiteCommandee + " unités commandées. "
                        + quantiteManquante + " unités en attente. Stock minimum réservé: "
                        + stockMinimum
                : "Commande complète : " + quantiteCommandee
                        + " unités commandées. Stock minimum réservé: " + stockMinimum;

        return new CommandeResult(
                true,
                quantiteManquante > 0 ? "partiel" : "complet",
                mapper.toCommande(commande, stockLookup(fm.getFournisseur().getId())),
                quantiteCommandee, quantiteManquante,
                stockAvant, nouveauStock, stockMinimum,
                message,
                List.of());
    }

    @Transactional(readOnly = true)
    public CommandeResponse getDetail(UUID commandeId) {
        CommandeFournisseur commande = commandeRepository.findById(commandeId)
                .orElseThrow(() -> new ResourceNotFoundException("commande", commandeId));
        return mapper.toCommande(commande,
                stockLookup(commande.getFournisseur().getId()));
    }

    /**
     * Create a supplier-only stock alert after an order drives the supplier
     * stock to its minimum (mirrors legacy {@code verifierEtatStock}).
     */
    private void verifierEtatStock(FournisseurMedicament fm, String nomMedicament) {
        int stockRestant = fm.getStockDisponible();
        int stockMinimum = fm.getStockMinimum();
        int seuilAlerte = fm.getSeuilReapprovisionnement();

        if (stockRestant <= stockMinimum) {
            Alerte alerte = new Alerte();
            alerte.setType(AlerteType.stock);
            alerte.setNiveau(AlerteNiveau.eleve);
            alerte.setMessage("STOCK MINIMUM ATTEINT : " + nomMedicament
                    + ". Stock restant: " + stockRestant + " unités (minimum: " + stockMinimum
                    + "). Veuillez réapprovisionner.");
            alerte.setEstLue(false);
            alerte.setDonneesConcernees(donneesFournisseur(fm, nomMedicament, stockRestant,
                    null, "stock_minimum_atteint"));
            alerteRepository.save(alerte);
            log.warn("Stock minimum atteint [medicament={}, stock_restant={}, stock_minimum={}]",
                    nomMedicament, stockRestant, stockMinimum);
        } else if (stockRestant <= seuilAlerte) {
            Alerte alerte = new Alerte();
            alerte.setType(AlerteType.stock);
            alerte.setNiveau(AlerteNiveau.moyen);
            alerte.setMessage("STOCK FAIBLE : " + nomMedicament + ". Stock restant: "
                    + stockRestant + " unités. Seuil d'alerte: " + seuilAlerte);
            alerte.setEstLue(false);
            alerte.setDonneesConcernees(donneesFournisseur(fm, nomMedicament, stockRestant,
                    seuilAlerte, "stock_faible"));
            alerteRepository.save(alerte);
            log.info("Stock faible [medicament={}, stock_restant={}, seuil_alerte={}]",
                    nomMedicament, stockRestant, seuilAlerte);
        }
    }

    private Map<String, Object> donneesFournisseur(FournisseurMedicament fm, String nomMedicament,
                                                   int stockRestant, Integer seuilAlerte,
                                                   String typeDetail) {
        Map<String, Object> donnees = new LinkedHashMap<>();
        donnees.put("fournisseur_id", fm.getFournisseur().getId().toString());
        donnees.put("medicament_id", fm.getMedicament().getId().toString());
        donnees.put("medicament_nom", nomMedicament);
        donnees.put("stock_restant", stockRestant);
        donnees.put("stock_minimum", fm.getStockMinimum());
        if (seuilAlerte != null) {
            donnees.put("seuil_alerte", seuilAlerte);
        }
        donnees.put("for_fournisseur", true);
        donnees.put("type_detail", typeDetail);
        return donnees;
    }

    private List<FournisseurMedicamentResponse> alternatifs(UUID medicamentId, int quantite) {
        return fournisseurMedicamentRepository.findAlternatifs(medicamentId, quantite).stream()
                .map(mapper::toFmResponse)
                .toList();
    }

    private UUID pharmacieIdOf(UUID userId) {
        return userRepository.findById(userId)
                .map(User::getPharmacie)
                .flatMap(p -> Optional.ofNullable(p.getId()))
                .orElse(null);
    }

    private Function<UUID, Optional<FournisseurMedicament>> stockLookup(UUID fournisseurId) {
        return medicamentId -> fournisseurMedicamentRepository
                .findByFournisseurIdAndMedicamentId(fournisseurId, medicamentId);
    }

    private String genererNumero() {
        return "CMD-" + LocalDateTime.now().format(
                DateTimeFormatter.ofPattern("yyyyMMddHHmmss")) + "-"
                + ThreadLocalRandom.current().nextInt(100, 1000);
    }

    private String nom(FournisseurMedicament fm) {
        String nom = fm.getMedicament().getNomCommercialFr();
        return nom != null ? nom : "Médicament";
    }
}